ui/webui/src/components/app.jsx

```diff
@@ -43,14 +43,73 @@ import { usePageLocation } from "hooks";
 
 const _ = cockpit.gettext;
 
+const getStep = ({ address, currentStepId, onAddErrorNotification, stepNotification, stepsOrder, stepsVisited }) => {
+    const wrapWithContext = (children) => {
+        return (
+            <AddressContext.Provider value={address}>
+                <Stack hasGutter>
+                    {stepNotification &&
+                     (stepNotification.step === currentStepId) &&
+                     <Alert
+                       isInline
+                       title={stepNotification.message}
+                       variant="danger"
+                     />}
+                    {children}
+                </Stack>
+            </AddressContext.Provider>
+        );
+    };
+
+    return stepsOrder.map((s, idx) => {
+        const Renderer = s.c;
+
+        return ({
+            id: s.id,
+            name: s.l,
+            component: wrapWithContext(<Renderer onAddErrorNotification={onAddErrorNotification} />),
+            stepNavItemProps: { id: s.id },
+            canJumpTo: idx === 0 ? currentStepId == s.id : stepsVisited.includes(s.id),
+            isFinishedStep: idx === stepsOrder.length - 1
+        });
+    });
+};
+
 export const Application = () => {
+    const stepsOrder = [
+        {
+            c: InstallationLanguage,
```

Comment:
Even if its some extra typing I would recommend to use full names for the keys, eq. "component", "label". Otherwise I think it could be a little too magic elsewhere in the code & not clear where the names come from. :)

Reply:
Agreed :)